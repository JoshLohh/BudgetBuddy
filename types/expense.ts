export interface Expense {
  amount: number | string;
  paidBy: string;
  splitBetween: string[];   // user IDs
  splitType: 'equal' | 'exact' | 'percentage';
  customSplit?: string;   // JSON string: { [userId]: number }
  description?: string;
  groupId?: string;
  category?: string;
  createdAt?: string;
  $id?: string;
}