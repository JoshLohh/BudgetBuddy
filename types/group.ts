export interface Group {
  id: string;
  title: string;
  description?: string;
  members: string[];
  createdBy: string;
}

export interface CreateGroupInput {
  title: string;
  description?: string;
  // Add more fields as needed
}

export interface MemberProfile {
  userId: string;
  username: string;
  avatar: string | null;
}