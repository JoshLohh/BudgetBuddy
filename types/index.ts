export * from './user';
export * from './group';
export * from './expense';
export * from './settlement';
