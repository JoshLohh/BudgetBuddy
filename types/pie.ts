export interface CategoryPieData {
  value: number;
  color: string;
  label: string;
  icon: string;
}

export interface MemberPieData {
  value: number;
  color: string;
  label: string;
  avatar: string | null;
}

export type PieDataItem = CategoryPieData | MemberPieData;
