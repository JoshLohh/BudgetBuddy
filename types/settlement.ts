export interface Settlement {
  groupId: string;
  from: string;
  to: string;
  amount: number;
  $createdAt: string;
  $id: string;
}

export type Balances = Record<string, number>;
