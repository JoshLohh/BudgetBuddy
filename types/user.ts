export interface User {
  $id: string;
  email: string;
  username?: string;
  avatar?: string | null;
  bio?: string;
}

export interface UserProfile {
  userId: string;
  username: string;
  avatar: string | null;
  email?: string;
  bio?: string;
}
