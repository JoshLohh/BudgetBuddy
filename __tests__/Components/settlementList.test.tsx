import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import { Alert } from 'react-native';
import SettlementList from '../../components/settlementList';
import { Settlement } from '@/types';
import { ReactTestInstance } from 'react-test-renderer';

describe('SettlementList', () => {
  const userId = 'user1';

  const mockGetUsername = (id: string) => `User-${id}`;

  const mockSettleUp: (from: string, to: string, amount: number) => Promise<void> =
  jest.fn().mockResolvedValue(undefined);

  const baseSettlements: Settlement[] = [
    { from: 'user2', to: 'user1', amount: 10, groupId: 'g1', $createdAt: '', $id: '1' },
    { from: 'user1', to: 'user3', amount: 5, groupId: 'g1', $createdAt: '', $id: '2' },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders "you are owed" statement with proper color', async () => {
    const { getByText } = render(
      <SettlementList
        currentUserId="user1"
        suggestedSettlements={baseSettlements}
        getUsername={mockGetUsername}
        settleUp={mockSettleUp}
      />
    );

    await waitFor(() => {
      expect(getByText('You are owed $5.00')).toBeTruthy();
    });
  });

  it('renders "you owe" statement when net balance is negative', async () => {
    const negativeSettlements: Settlement[] = [
      { from: 'user1', to: 'user2', amount: 50, groupId: 'g1', $createdAt: '', $id: '1' },
    ];

    const { getByText } = render(
      <SettlementList
        currentUserId="user1"
        suggestedSettlements={negativeSettlements}
        getUsername={mockGetUsername}
        settleUp={mockSettleUp}
      />
    );

    await waitFor(() => {
      expect(getByText('You owe $50.00')).toBeTruthy();
    });
  });

  it('renders "no settlements needed" when list is empty', async () => {
    const { getByText } = render(
      <SettlementList
        currentUserId="userX"
        suggestedSettlements={[]}
        getUsername={mockGetUsername}
        settleUp={mockSettleUp}
      />
    );

    await waitFor(() => {
      expect(getByText('No settlements needed.')).toBeTruthy();
    });
  });

  it('renders all settlements for the current user', async () => {
    const { getByText, getAllByText } = render(
      <SettlementList
        currentUserId="user1"
        suggestedSettlements={baseSettlements}
        getUsername={mockGetUsername}
        settleUp={mockSettleUp}
      />
    );
  
    await waitFor(() => {
      expect(getByText('$10.00')).toBeTruthy();
      expect(getByText('$5.00')).toBeTruthy();
      expect(getByText(/User-user2/)).toBeTruthy();
      expect(getAllByText(/User-user1/).length).toBeGreaterThanOrEqual(2);
      expect(getByText(/User-user3/)).toBeTruthy();
    });
  });

  it('calls settleUp after confirm in Alert', async () => {
    jest.spyOn(Alert, 'alert').mockImplementation((title, message, buttons) => {
      const confirmButton = buttons?.find(button => button.text === 'Confirm');
      if (confirmButton && confirmButton.onPress) {
        act(() => {
          if (confirmButton.onPress) {
            confirmButton.onPress();
          }
        });
      }
    });

    const { getAllByText } = render(
      <SettlementList
        currentUserId="user1"
        suggestedSettlements={baseSettlements}
        getUsername={mockGetUsername}
        settleUp={mockSettleUp}
      />
    );

    const buttons = getAllByText('Settle Up');
    fireEvent.press(buttons[1]); // The one where user1 is "from"

    await waitFor(() => {
      expect(mockSettleUp).toHaveBeenCalledWith('user1', 'user3', 5);
    });
  });

  it('disables Settle Up button during processing', async () => {
    const longSettleUp: (from: string, to: string, amount: number) => Promise<void> =
      (from, to, amount) => new Promise(resolve => setTimeout(resolve, 100));
  
    const { getAllByText, getAllByRole } = render(
      <SettlementList
        currentUserId="user1"
        suggestedSettlements={baseSettlements}
        getUsername={mockGetUsername}
        settleUp={longSettleUp}
      />
    );
  
    const buttons = getAllByText('Settle Up');
  
    await act(async () => {
      fireEvent.press(buttons[1]); // e.g., settlement from user1 â†’ user3
    });
  
    const pressables = getAllByRole('button');
  
    const anyButtonDisabled = pressables.some(
      node => node.props.accessibilityState?.disabled === true
    );
  
    expect(anyButtonDisabled).toBe(true);
  });
  
  //Snapshot test
  it('matches snapshot with typical settlement list', () => {
    const settlements = [
      { from: 'user2', to: 'user1', amount: 10, groupId: 'g1', $createdAt: '', $id: '1' },
      { from: 'user1', to: 'user3', amount: 5, groupId: 'g1', $createdAt: '', $id: '2' },
    ];
    const { toJSON } = render(
      <SettlementList
        currentUserId="user1"
        suggestedSettlements={settlements}
        getUsername={(id) => `User-${id}`}
        settleUp={jest.fn().mockResolvedValue(undefined)}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });
  
});

/**
 * Summary:
 * - Ensures correct rendering logic based on user settlements and roles.
 * - Uses waitFor and act to avoid "not wrapped in act(...)" errors.
 * - Mocks Alert.confirm behavior to simulate user action.
 * - Verifies styling/outcome text (e.g., 'You are owed') and interactive logic.
 */
