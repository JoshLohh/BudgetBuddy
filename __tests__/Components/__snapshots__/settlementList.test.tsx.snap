// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettlementList matches snapshot with typical settlement list 1`] = `
<View
  style={
    {
      "marginTop": 12,
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#11181C",
        },
        undefined,
        undefined,
        undefined,
        {
          "fontSize": 18,
          "fontWeight": "bold",
        },
        undefined,
        undefined,
        {
          "fontWeight": "bold",
          "marginBottom": 8,
        },
      ]
    }
  >
    Suggested Settlements
  </Text>
  <Text
    style={
      [
        {
          "color": "#11181C",
        },
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        {
          "color": "#00cc66",
          "fontWeight": "bold",
        },
      ]
    }
  >
    You are owed $5.00
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 8,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "flex": 1,
            "fontWeight": "bold",
          },
        ]
      }
    >
      User-user2
       
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#1976d2",
              "fontSize": 16,
            },
            undefined,
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
       
      User-user1
    </Text>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "color": "#1e88e5",
            "fontWeight": "bold",
            "width": 80,
          },
        ]
      }
    >
      $
      10.00
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderRadius": 6,
            "marginVertical": 10,
            "padding": 18,
          },
          {
            "backgroundColor": "#0a7ea4",
          },
          false,
          {
            "marginLeft": 10,
            "minWidth": 90,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "color": "#fff",
            },
          ]
        }
      >
        Settle Up
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 8,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "flex": 1,
            "fontWeight": "bold",
          },
        ]
      }
    >
      User-user1
       
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#1976d2",
              "fontSize": 16,
            },
            undefined,
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
       
      User-user3
    </Text>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "color": "#1e88e5",
            "fontWeight": "bold",
            "width": 80,
          },
        ]
      }
    >
      $
      5.00
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderRadius": 6,
            "marginVertical": 10,
            "padding": 18,
          },
          {
            "backgroundColor": "#0a7ea4",
          },
          false,
          {
            "marginLeft": 10,
            "minWidth": 90,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "color": "#fff",
            },
          ]
        }
      >
        Settle Up
      </Text>
    </View>
  </View>
</View>
`;
